### PR TITLE
A couple of ASTScope-related cleanups

### PR DESCRIPTION
**Pull request by slavapestov**
More groundwork for disabling parse-time lookup.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

File: include/swift/AST/ASTScope.h [hunk 1]
```diff
@@ -1945,17 +1945,55 @@ class CaseStmtScope final : public AbstractStmtScope {
 private:
   void expandAScopeThatDoesNotCreateANewInsertionPoint(ScopeCreator &);
 
+public:
+  std::string getClassName() const override;
+  Stmt *getStmt() const override { return stmt; }
+};
+
+class CaseLabelItemScope final : public ASTScopeImpl {
```
Comment (davidungar) [inline include/swift/AST/ASTScope.h:1953]: Looks good. How about a comment for the two new scopes explaining what they model? Maybe a tiny Swift code example??

File: lib/AST/ASTScopeCreation.cpp [hunk 2]
```diff
@@ -1486,10 +1488,21 @@ void ForEachPatternScope::expandAScopeThatDoesNotCreateANewInsertionPoint(
...
-  for (auto &caseItem : stmt->getMutableCaseLabelItems())
-    scopeCreator.addToScopeTree(caseItem.getGuardExpr(), this);
+  for (auto &item : stmt->getCaseLabelItems())
+    scopeCreator.constructExpandAndInsertUncheckable<CaseLabelItemScope>(
+      this, item);
 
-  // Add a child for the case body.
+  scopeCreator.constructExpandAndInsertUncheckable<CaseStmtBodyScope>(
+      this, stmt);
+}
```
Comment (davidungar) [inline lib/AST/ASTScopeCreation.cpp:1497]: Nice.

File: lib/AST/ASTScopeLookup.cpp [hunk 3]
```diff
@@ -422,12 +422,18 @@ bool ForEachPatternScope::lookupLocalsOrMembers(ArrayRef<const ASTScopeImpl *>,
...
+  return lookupLocalBindingsInPattern(
+      item.getPattern(), DeclVisibilityKind::LocalVariable, consumer);
+}
+
+bool CaseStmtBodyScope::lookupLocalsOrMembers(ArrayRef<const ASTScopeImpl *>,
+                                              DeclConsumer consumer) const {
+  for (auto *var : stmt->getCaseBodyVariablesOrEmptyArray())
+    if (consumer.consume({var}, DeclVisibilityKind::LocalVariable))
+        return true;
+
```
Comment (davidungar) [inline lib/AST/ASTScopeLookup.cpp:436]: Looks good.

File: lib/AST/ASTScopeLookup.cpp [hunk 4]
```diff
@@ -554,7 +560,7 @@ bool ConditionalClausePatternUseScope::lookupLocalsOrMembers(
       pattern, DeclVisibilityKind::LocalVariable, consumer);
 }
 
-bool ASTScopeImpl::lookupLocalBindingsInPattern(Pattern *p,
+bool ASTScopeImpl::lookupLocalBindingsInPattern(const Pattern *p,
```
Comment (davidungar) [inline lib/AST/ASTScopeLookup.cpp:563]: I'm grinning at that `const`!

File: lib/AST/ASTScopeSourceRange.cpp [hunk 5]
```diff
@@ -408,20 +408,13 @@ SourceRange ForEachPatternScope::getSourceRangeOfThisASTNode(
 }
 
 SourceRange
-CaseStmtScope::getSourceRangeOfThisASTNode(const bool omitAssertions) const {
-  // The scope of the case statement begins at the first guard expression,
-  // if there is one, and extends to the end of the body.
-  // FIXME: Figure out what to do about multiple pattern bindings. We might
-  // want a more restrictive rule in those cases.
```
Comment (davidungar) [inline lib/AST/ASTScopeSourceRange.cpp:415]: Is there some new comment that belongs somewhere to replace this one?

File: lib/AST/ASTScopeSourceRange.cpp [hunk 6]
```diff
@@ -408,20 +408,13 @@ SourceRange ForEachPatternScope::getSourceRangeOfThisASTNode(
...
+  return item.getSourceRange();
+}
 
-  // Otherwise, it covers the body.
-  return stmt->getBody()
-      ->getSourceRange(); // The scope of the case statement begins
+SourceRange
+CaseStmtBodyScope::getSourceRangeOfThisASTNode(const bool omitAssertions) const {
+  return stmt->getBody()->getSourceRange();
 }
```
Comment (davidungar) [inline lib/AST/ASTScopeSourceRange.cpp:418]: The disappearance of this function is a good smell.

**Review by davidungar — APPROVED** [review]
Very nice! Maybe just add a few comments (as noted). Thanks. Great to see the ASTScope stuff extend so gracefully.

File: include/swift/AST/ASTScope.h (hunk 1 above)
Comment (slavapestov) [reply include/swift/AST/ASTScope.h:1953]: Done.

File: lib/AST/ASTScopeSourceRange.cpp (hunk 5 above)
Comment (slavapestov) [reply lib/AST/ASTScopeSourceRange.cpp:415]: I've added a comment to CaseStmtScope that I think explains everything.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

File: include/swift/AST/ASTScope.h [hunk 7]
```diff
@@ -1933,6 +1933,42 @@ class ForEachPatternScope final : public ASTScopeImpl {
...
+/// | +----------------------------------------------------------------
+/// | | CaseStmtBodyScope:
+/// | +----------------------------------------------------------------
+/// | | {
+/// | |    ... x, y  <-- body sees "joined" 'x'/'y' created by parser
+/// | | }
+/// | +----------------------------------------------------------------
+/// |
+/// +------------------------------------------------------------------
+
```
Comment (davidungar) [inline include/swift/AST/ASTScope.h:1971]: Fantastic! Thank you.

File: include/swift/AST/ASTScope.h [hunk 8]
```diff
@@ -1945,17 +1981,66 @@ class CaseStmtScope final : public AbstractStmtScope {
...
 
+public:
+  std::string getClassName() const override;
+  Stmt *getStmt() const override { return stmt; }
+};
+
+/// The scope used for the guard expression in a case statement. Any
+/// variables bound by the case label item's pattern are visible in
+/// this scope.
+class CaseLabelItemScope final : public ASTScopeImpl {
```
Comment (davidungar) [inline include/swift/AST/ASTScope.h:1992]: Great! Thank you

File: include/swift/AST/ASTScope.h [hunk 9]
```diff
@@ -1945,17 +1981,66 @@ class CaseStmtScope final : public AbstractStmtScope {
...
 };
 
+/// The scope used for the body of a 'case' statement.
+///
+/// If the 'case' statement has multiple case label items, each label
+/// item's pattern must bind the same variables; the parser creates
+/// "fake" variables to represent the join of the variables bound by
+/// each pattern.
+///
+/// These "fake" variables are visible in the 'case' statement body.
```
Comment (davidungar) [inline include/swift/AST/ASTScope.h:2021]: Great "why" comment!

**Review by davidungar — APPROVED** [review]
Looks even better now!

**Comment by slavapestov** [comment]
@swift-ci Please smoke test macOS

**Pull request merged**